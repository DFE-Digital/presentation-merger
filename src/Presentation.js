/** @module Presentation */
import { get } from 'shvl';
import { namespaces, uuid } from './utils';

/**
 * Class prepresenting `content.xml` in the presentation document
 */
export default class Presentation {
  /**
   * Create a presentation content representation.
   * @param {object} presentation JSON representation of content.xml
   * @param {string} id unique identifier for the presentation
   */
  constructor(presentation, id) {
    this.id = id;
    if (id === undefined) {
      this.id = uuid();
    }
    this.data = presentation;
    this._uniqueStyleIDs();
    this.presentation = this.data;
    this.namespaces = namespaces(presentation['office:document-content']);
  }

  _uniqueStyleIDs() {
    [
      'office:document-content.office:automatic-styles.style:style',
      'office:document-content.office:automatic-styles.text:list-style',
    ].forEach(key => {
      this.data = this._renameStyleKeys(this.data, key);
    });
  }

  _renameStyleKeys(document, key) {
    let styles = get(document, key);
    let keys = new Set();
    styles.forEach(style => {
      Object.entries(style).forEach(([key, value]) => {
        if (key.endsWith(':name')) {
          keys.add(value);
        }
      });
    });
    let str = JSON.stringify(document);
    keys.forEach(key => {
      let newId = `${this.id}-${key}`;
      str = str.replace(new RegExp(key, 'g'), newId);
    });
    return JSON.parse(str);
  }
}
