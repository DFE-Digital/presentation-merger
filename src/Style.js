/** @module Style */
import XMLFileBase from './XMLFileBase';
// import { extractArray } from './utils';

/**
 * Class prepresenting `style.xml` in the presentation document
 */
export default class Style extends XMLFileBase {
  /**
   * Create a style representation
   * @param {string} xml XML content
   * @param {array} manifest The document manifest of image assets
   * @param {string} id unique identifier for the presentation
   */
  constructor(xml, manifest, id) {
    super(xml, id, 'office:document-styles', manifest);
  }

  /*
  get arrayNotation() {
    return [
      'style:handout-master',
      'style:master-page',
      'style:style',
      'style:page-layout',
      'style:presentation-page-layout',
      'style:default-style',
      'draw:layer',
      'draw:stroke-dash',
      'draw:marker',
      'draw:fill-image',
      'draw:hatch',
      'draw:gradient',
      'draw:frame',
      'draw:custom-shape'
    ];
  }
  */

  get keys() {
    return [
      'office:document-styles.office:font-face-decls.style:font-face',
      'office:document-styles.office:styles.draw:gradient',
      'office:document-styles.office:styles.draw:hatch',
      'office:document-styles.office:styles.draw:fill-image',
      'office:document-styles.office:styles.draw:marker',
      'office:document-styles.office:styles.draw:stroke-dash',
      'office:document-styles.office:styles.style:default-style',
      'office:document-styles.office:styles.style:style',
      /**
       * @warning `presentation-page-layout` causes OpenOffice to crash when opening the document
       */
      // 'office:document-styles.office:styles.style:presentation-page-layout',
      // 'office:document-styles.office:automatic-styles.style:page-layout',
      'office:document-styles.office:automatic-styles.style:style',
      'office:document-styles.office:automatic-styles.text:list-style',
      // 'office:document-styles.office:master-styles.draw:layer-set.draw:layer',
      // 'office:document-styles.office:master-styles.style:handout-master',
      'office:document-styles.office:master-styles.style:master-page'
    ];
  }

  // eslint-disable-next-line max-lines-per-function
  changeKeyReferences(content) {
    // let arr = extractArray(
    //   content,
    //   'office:document-styles.office:automatic-styles.style:style'
    // );
    // let mp = extractArray(
    //   content,
    //   'office:document-styles.office:master-styles.style:master-page'
    // );
    // mp.forEach(e => {
    //   delete e['presentation:notes'];
    //   delete e['draw:custom-shape']; // @temp
    //   extractArray(e, 'draw:custom-shape').forEach(s => {
    //     if (s['@_presentation:style-name']) {
    //       s['@_presentation:style-name'] =
    //         this.id + '_' + s['@_presentation:style-name'];
    //     }
    //   });
    // });
    // arr.forEach(e => {
    //   if (e['@_style:family'] === 'presentation') {
    //     e['@_style:name'] = this.id + '_' + e['@_style:name'];
    //   }
    // });

    return content;
  }
}
