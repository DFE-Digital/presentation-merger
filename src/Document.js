import { get, set } from "shvl";
import JSZip from "jszip";
import decompress from "decompress";
import EventEmitter from "events";
import { toJson, toXml } from "xml2json";
import { format } from "prettier";

import Presentation from "../src/Presentation";

export default class Document extends EventEmitter {
  constructor() {
    super({ captureRejections: true });
    this.slides = [];
    this.files = [];
    this.manifestFiles = [
      {
        mimeType: "application/vnd.oasis.opendocument.presentation",
        path: "/",
        version: "1.2"
      },
      {
        mimeType: "text/xml",
        path: "content.xml"
      }
    ];
    this.counter = 0;
    this.doc = {
      "office:document-content": {}
    };
  }

  async mergeFile(file) {
    const files = await decompress(file);
    const manifest = files.find(f => f.path === "META-INF/manifest.xml");
    let json = JSON.parse(toJson(manifest.data));
    const manifestFiles = get(json, "manifest:manifest.manifest:file-entry");
    const content = files.find(f => f.path === "content.xml");
    let pres = new Presentation(JSON.parse(toJson(content.data)));
    this.merge(pres);
    manifestFiles.forEach(manifestFile => {
      if (get(manifestFile, "manifest:media-type").startsWith("image/")) {
        const file = files.find(
          f => f.path === manifestFile["manifest:full-path"]
        );
        file.path = `Presentation${this.counter}/${file.path}`;
        this.files.push(file);
        this.manifestFiles.push({
          mimeType: manifestFile["manifest:media-type"],
          path: file.path
        });
      }
    });
    this.counter = this.counter + 1;
  }

  merge(pres) {
    if (!(pres instanceof Presentation)) {
      throw new TypeError(
        `expected Presentation class got ${pres.constructor.name}`
      );
    }
    this.slides = this.slides.concat(pres.slides);
    let slides = this.slides.map(slidesDocument => slidesDocument.content);
    let styles = this.slides
      .map(slidesDocument => slidesDocument.styles)
      .flat();

    Object.assign(this.doc["office:document-content"], pres.namespaces);
    set(
      this.doc,
      [
        "office:document-content",
        "office:body",
        "office:presentation",
        "draw:page"
      ],
      slides
    );

    set(
      this.doc,
      ["office:document-content", "office:automatic-styles", "style:style"],
      styles
    );
  }

  pipe(stream) {
    let zip = new JSZip();
    this.files.forEach(file => {
      zip.file(file.path, file.data);
    });
    zip.file("mimetype", "application/vnd.oasis.opendocument.presentation");
    zip.file("content.xml", this.toFormattedXML(this.doc));
    zip.file("META-INF/manifest.xml", this.toFormattedXML(this.manifest));
    let promise = new Promise((resolve, reject) => {
      zip
        .generateNodeStream({
          type: "nodebuffer",
          streamFiles: true,
          compression: "DEFLATE",
          compressionOptions: {
            level: 9
          }
        })
        .pipe(stream)
        .on("finish", () => {
          resolve();
          this.emit("end");
        })
        .on("error", err => {
          reject(err);
          this.emit("error", err);
        });
    });
    return promise;
  }

  get manifest() {
    let output = {
      "manifest:manifest": {
        "xmlns:manifest": "urn:oasis:names:tc:opendocument:xmlns:manifest:1.0",
        "manifest:version": "1.2",
        "manifest:file-entry": []
      }
    };
    output["manifest:manifest"]["manifest:file-entry"] = this.manifestFiles.map(
      file => {
        let out = {
          "manifest:full-path": file.path,
          "manifest:media-type": file.mimeType
        };
        if (file.version) {
          out["manifest:version"] = file.version;
        }
        return out;
      }
    );
    return output;
  }

  toFormattedXML(object) {
    return format(toXml(object), {
      xmlSelfClosingSpace: true,
      xmlWhitespaceSensitivity: "ignore",
      parser: "xml"
    });
  }
}
