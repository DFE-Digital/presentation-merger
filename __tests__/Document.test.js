import fs from 'fs';
import path from 'path';
import { get } from 'shvl';
import { toJson } from 'xml2json';
import decompress from 'decompress';
import Presentation from '../src/Presentation';
import Document from '../src/Document';
import Style from '../src/Style';

const fixture1 = JSON.parse(
  fs.readFileSync(path.join(__dirname, '__fixtures__/file1.json'))
);

const fixture2 = JSON.parse(
  fs.readFileSync(path.join(__dirname, '__fixtures__/file2.json'))
);

describe('Document', () => {
  let subject;
  let style;
  let presentation;
  beforeEach(async () => {
    const files = await decompress(
      path.join(__dirname, '__fixtures__/file1.odp')
    );
    const masterStylesFile = files.find(f => f.path === 'styles.xml');
    const contentFile = files.find(f => f.path === 'content.xml');
    presentation = new Presentation(JSON.parse(toJson(contentFile.data)));
    style = new Style(JSON.parse(toJson(masterStylesFile.data)), presentation);
    subject = new Document();
  });
  
  describe('.mergeFile', () => {
    it('prepares files to merge', async () => {
      let doc = new Document();
      await doc.mergeFile(path.join(__dirname, '__fixtures__/file1.odp'));
      await doc.mergeFile(path.join(__dirname, '__fixtures__/file2.odp'));
      let stream = fs.createWriteStream(
        path.join(__dirname, '__fixtures__/out1.odp'),
        { flags: 'w' }
      );
      doc.pipe(stream);
      return new Promise(done => {
        doc.on('end', async () => {
          const files = await decompress(
            path.join(__dirname, '__fixtures__/out1.odp'),
            'out'
          );
          let filePaths = files.map(f => f.path);

          expect(filePaths).toContainEqual('content.xml');
          expect(filePaths).toContainEqual(
            'Presentation0/Pictures/100000000000032000000258E080B12F.jpg'
          );

          expect(filePaths).toContainEqual(
            'Presentation0/Pictures/100002010000028200000040779DD47E.png'
          );

          expect(filePaths).toContainEqual(
            'Presentation0/Pictures/1000000000000040000000400142E835.png'
          );
          done();
        });
        doc.on('error', err => {
          throw err;
        });
      });
    });
  });

  describe('.merge', () => {
    it('merges multiple presentations', () => {
      let doc = new Document();
      let fixtures = [fixture1, fixture2];
      let presentations = [];

      fixtures.forEach(fixture => {
        presentations.push(new Presentation(fixture));
      });

      presentations.forEach(pres => {
        doc.merge(pres);
      });

      let slides = get(
        doc.doc,
        'office:document-content.office:body.office:presentation.draw:page'
      );

      fixtures.forEach(fixture => {
        let fixtureSlides = get(
          fixture,
          'office:document-content.office:body.office:presentation.draw:page'
        );
        fixtureSlides.forEach(slide => {
          expect(slides).toContainEqual(slide);
        });
      });
    });
  });

  describe('manifest', () => {
    it('returns the correct formatting for mimetype', async () => {
      expect(subject.manifest).toMatchSnapshot();
      await subject.mergeFile(path.join(__dirname, '__fixtures__/file1.odp'));
      expect(subject.manifest).toMatchSnapshot();
    });
  });

  describe('mergeStyles', () => {
    it('merges style content from the presentation into the document style', () => {
      subject.mergeStyles(style, presentation);
      expect(subject.stylesDoc).toMatchSnapshot();
    });
  });
});
