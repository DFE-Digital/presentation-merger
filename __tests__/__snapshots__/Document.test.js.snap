// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document manifest returns the correct formatting for mimetype 1`] = `
Array [
  Object {
    "manifest:full-path": "/",
    "manifest:media-type": "application/vnd.oasis.opendocument.presentation",
    "manifest:version": "1.2",
  },
  Object {
    "manifest:full-path": "content.xml",
    "manifest:media-type": "text/xml",
  },
]
`;

exports[`Document manifest returns the correct formatting for mimetype 2`] = `
Array [
  Object {
    "manifest:full-path": "/",
    "manifest:media-type": "application/vnd.oasis.opendocument.presentation",
    "manifest:version": "1.2",
  },
  Object {
    "manifest:full-path": "content.xml",
    "manifest:media-type": "text/xml",
  },
  Object {
    "manifest:full-path": "Presentation0/Thumbnails/thumbnail.png",
    "manifest:media-type": "image/png",
  },
  Object {
    "manifest:full-path": "Presentation0/Pictures/100000000000032000000258E080B12F.jpg",
    "manifest:media-type": "image/jpeg",
  },
  Object {
    "manifest:full-path": "Presentation0/Pictures/100002010000028200000040779DD47E.png",
    "manifest:media-type": "image/png",
  },
  Object {
    "manifest:full-path": "Presentation0/Pictures/1000000000000040000000400142E835.png",
    "manifest:media-type": "image/png",
  },
]
`;
